File: arrays-hashing/is-valid-sudoku.py

from typing import List


def isValidSudoku(board: List[List[str]]) -> bool:
    rows = [set() for _ in range(9)]
    columns = [set() for _ in range(9)]
    boxes = [set() for _ in range(9)]

    for r in range(9):
        for c in range(9):
            b = (r//3) * 3 + (c//3)
            ch = board[r][c]
            if ch == ".": 
                continue

            if ch in rows[r] or ch in columns[c] or ch in boxes[b]:
                return False
            else:
                rows[r].add(ch)
                columns[c].add(ch)
                boxes[b].add(ch)
    
    return True

board = [["1","2",".",".","3",".",".",".","."],["4",".",".","5",".",".",".",".","."],[".","9","8",".",".",".",".",".","3"],["5",".",".",".","6",".",".",".","4"],[".",".",".","8",".","3",".",".","5"],["7",".",".",".","2",".",".",".","6"],[".",".",".",".",".",".","2",".","."],[".",".",".","4","1","9",".",".","8"],[".",".",".",".","8",".",".","7","9"]]
invalid_board = [["1","2",".",".","3",".",".","1","."],["4",".",".","5",".",".",".",".","."],[".","9","8",".",".",".",".",".","3"],["5",".",".",".","6",".",".",".","4"],[".",".",".","8",".","3",".",".","5"],["7",".",".",".","2",".",".",".","6"],[".",".",".",".",".",".","2",".","."],[".",".",".","4","1","9",".",".","8"],[".",".",".",".","8",".",".","7","9"]]

print(f"Valid board: {isValidSudoku(board)}")
print(f"Invalid board: {isValidSudoku(invalid_board)}")

File: arrays-hashing/is-valid-sudoku.js
class Solution {
  /**
   * @param {character[][]} board
   * @return {boolean}
   */
  isValidSudoku(board) {
    // for each row, is row[i] unique
    // for each column is col[j] unique
    // for each box is (row / 3) * 3 + (col / 3) unique

    // brute force approach would be for each cell scan the entire row, column, and box for duplicates

    // we could define a board using Maps and iterate through the board placing each item in the Map
    // if double entry encountered, return false

    // board is 9x9, cells are '1'..'9' or '.'
    const rows = Array.from({ length: 9 }, () => new Set());
    const cols = Array.from({ length: 9 }, () => new Set());
    const boxes = Array.from({ length: 9 }, () => new Set());

    for (let r = 0; r < 9; r++) {
      for (let c = 0; c < 9; c++) {
        const ch = board[r][c];
        if (ch === ".") continue; // skip empties

        const box = Math.floor(r / 3) * 3 + Math.floor(c / 3); // 0..8

        if (rows[r].has(ch) || cols[c].has(ch) || boxes[box].has(ch)) {
          return false; // duplicate found
        }
        rows[r].add(ch);
        cols[c].add(ch);
        boxes[box].add(ch);
      }
    }

    return true;
  }
}
